File: epi_judge_java/epi/IntervalsUnion.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.EpiUserType;
import epi.test_framework.GenericTest;
import epi.test_framework.TimedExecutor;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public class IntervalsUnion {

    public static List<Interval> unionOfIntervals(List<Interval> intervals) {
        List<Interval> ans = new ArrayList<>();
        intervals.sort((i1, i2) -> {
            if (i1.left.val == i2.left.val) {
                return i1.left.isClosed ? -1 : 1;
            }
            return i1.left.val - i2.left.val;
        });

        Interval.Endpoint currLeft = intervals.get(0).left;
        Interval.Endpoint currRight = intervals.get(0).right;
        for (int i = 1; i < intervals.size(); i++) {
            var next = intervals.get(i);
            if (next.left.val > currRight.val
                    || (next.left.val == currRight.val && !next.left.isClosed && !currRight.isClosed)) {
                ans.add(new Interval(currLeft, currRight));
                currLeft = next.left;
                currRight = next.right;
            } else {
                if (next.left.val < currLeft.val || (next.left.val == currLeft.val && next.left.isClosed)) {
                    currLeft = next.left;
                }
                if (next.right.val > currRight.val || (next.right.val == currRight.val && next.right.isClosed)) {
                    currRight = next.right;
                }
            }
        }
        ans.add(new Interval(currLeft, currRight));
        return ans;
    }

    public static class Interval {
        public Endpoint left = new Endpoint();
        public Endpoint right = new Endpoint();

        public Interval() {
        }

        public Interval(Endpoint left, Endpoint right) {
            this.left = left;
            this.right = right;
        }

        private static class Endpoint {
            public boolean isClosed;
            public int val;
        }
    }

    @EpiUserType(
            ctorParams = {int.class, boolean.class, int.class, boolean.class})
    public static class FlatInterval {
        int leftVal;
        boolean leftIsClosed;
        int rightVal;
        boolean rightIsClosed;

        public FlatInterval(int leftVal, boolean leftIsClosed, int rightVal,
                            boolean rightIsClosed) {
            this.leftVal = leftVal;
            this.leftIsClosed = leftIsClosed;
            this.rightVal = rightVal;
            this.rightIsClosed = rightIsClosed;
        }

        public FlatInterval(Interval i) {
            if (i != null) {
                leftVal = i.left.val;
                leftIsClosed = i.left.isClosed;
                rightVal = i.right.val;
                rightIsClosed = i.right.isClosed;
            }
        }

        public Interval toInterval() {
            Interval i = new Interval();
            i.left.val = leftVal;
            i.left.isClosed = leftIsClosed;
            i.right.val = rightVal;
            i.right.isClosed = rightIsClosed;
            return i;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) {
                return true;
            }
            if (o == null || getClass() != o.getClass()) {
                return false;
            }

            FlatInterval that = (FlatInterval) o;

            if (leftVal != that.leftVal) {
                return false;
            }
            if (leftIsClosed != that.leftIsClosed) {
                return false;
            }
            if (rightVal != that.rightVal) {
                return false;
            }
            return rightIsClosed == that.rightIsClosed;
        }

        @Override
        public int hashCode() {
            int result = leftVal;
            result = 31 * result + (leftIsClosed ? 1 : 0);
            result = 31 * result + rightVal;
            result = 31 * result + (rightIsClosed ? 1 : 0);
            return result;
        }

        @Override
        public String toString() {
            return "" + (leftIsClosed ? "<" : "(") + leftVal + ", " + rightVal +
                    (rightIsClosed ? ">" : ")");
        }
    }

    @EpiTest(testDataFile = "intervals_union.tsv")
    public static List<FlatInterval>
    unionIntervalWrapper(TimedExecutor executor, List<FlatInterval> intervals)
            throws Exception {
        List<Interval> casted = new ArrayList<>(intervals.size());
        for (FlatInterval in : intervals) {
            casted.add(in.toInterval());
        }

        List<Interval> result = executor.run(() -> unionOfIntervals(casted));

        intervals = new ArrayList<>(result.size());
        for (Interval i : result) {
            intervals.add(new FlatInterval(i));
        }
        return intervals;
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "IntervalsUnion.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/intervals_union.cc
#include <vector>

#include "test_framework/generic_test.h"
#include "test_framework/serialization_traits.h"
#include "test_framework/timed_executor.h"
using namespace std;

struct Interval {
    struct Endpoint {
        bool is_closed;
        int val;
    };

    Endpoint left, right;
};

vector<Interval> UnionOfIntervals(vector<Interval> intervals) {
    if (intervals.empty()) return {};
    sort(intervals.begin(), intervals.end(), [](auto a, auto b) {
        if (a.left.val == b.left.val) return a.left.is_closed;
        return a.left.val < b.left.val;
    });

    vector<Interval> ans;
    Interval::Endpoint curr_left = intervals[0].left;
    Interval::Endpoint curr_right = intervals[0].right;
    for (int i = 1; i < intervals.size(); i++) {
        if (curr_right.val > intervals[i].left.val) {
            if (curr_right.val < intervals[i].right.val) {
                curr_right = intervals[i].right;
            } else if (curr_right.val == intervals[i].right.val && intervals[i].right.is_closed) {
                curr_right.is_closed = true;
            }
        } else if (curr_right.val == intervals[i].left.val && (curr_right.is_closed || intervals[i].left.is_closed)) {
            curr_right = intervals[i].right;
        } else {
            ans.push_back({curr_left, curr_right});
            curr_left = intervals[i].left;
            curr_right = intervals[i].right;
        }
    }
    ans.push_back({curr_left, curr_right});

    return ans;
}

struct FlatInterval {
    int left_val;
    bool left_is_closed;
    int right_val;
    bool right_is_closed;

    FlatInterval(int left_val, bool left_is_closed, int right_val,
                 bool right_is_closed)
        : left_val(left_val),
          left_is_closed(left_is_closed),
          right_val(right_val),
          right_is_closed(right_is_closed) {
    }

    explicit FlatInterval(Interval in)
        : left_val(in.left.val),
          left_is_closed(in.left.is_closed),
          right_val(in.right.val),
          right_is_closed(in.right.is_closed) {
    }

    operator Interval() const {
        return {{left_is_closed, left_val}, {right_is_closed, right_val}};
    }

    bool operator==(const FlatInterval &rhs) const {
        return std::tie(left_val, left_is_closed, right_val, right_is_closed) ==
               std::tie(rhs.left_val, rhs.left_is_closed, rhs.right_val,
                        rhs.right_is_closed);
    }
};

namespace test_framework {
    template<>
    struct SerializationTrait<FlatInterval>
            : UserSerTrait<FlatInterval, int, bool, int, bool> {
    };
} // namespace test_framework

std::ostream &operator<<(std::ostream &out, const FlatInterval &i) {
    return out << (i.left_is_closed ? '<' : '(') << i.left_val << ", "
           << i.right_val << (i.right_is_closed ? '>' : ')');
}

std::vector<FlatInterval> UnionOfIntervalsWrapper(
    TimedExecutor &executor, const std::vector<FlatInterval> &intervals) {
    std::vector<Interval> casted;
    for (const FlatInterval &i: intervals) {
        casted.push_back(static_cast<Interval>(i));
    }

    std::vector<Interval> result =
            executor.Run([&] { return UnionOfIntervals(casted); });

    return {begin(result), end(result)};
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"executor", "intervals"};
    return GenericTestMain(args, "intervals_union.cc", "intervals_union.tsv",
                           &UnionOfIntervalsWrapper, DefaultComparator{},
                           param_names);
}
